File: clients/ios/Lextures/Core/LMS/MasteryLogic.swift
import Foundation

enum MasteryLevel: String, Hashable {
    case mastered
    case developing
    case beginning
    case atRisk
    case notAssessed

    var labelKey: String {
        switch self {
        case .mastered: return "mobile.mastery.levelMastered"
        case .developing: return "mobile.mastery.levelDeveloping"
        case .beginning: return "mobile.mastery.levelBeginning"
        case .atRisk: return "mobile.mastery.levelAtRisk"
        case .notAssessed: return "mobile.mastery.levelNotAssessed"
        }
    }
}

struct MasteryConceptRow: Identifiable, Hashable {
    var id: String
    var name: String
    var score: Double?
    var assessed: Bool
    var level: MasteryLevel
}

enum MasteryLogic {
    /// Matches web `masteryLabel()` thresholds (0.8 / 0.6 / 0.4).
    static func level(score: Double?, assessed: Bool) -> MasteryLevel {
        guard assessed, let score else { return .notAssessed }
        if score >= 0.8 { return .mastered }
        if score >= 0.6 { return .developing }
        if score >= 0.4 { return .beginning }
        return .atRisk
    }

    static func rows(from row: StudentMasteryRow) -> [MasteryConceptRow] {
        let cellsByConcept = Dictionary(uniqueKeysWithValues: row.cells.map { ($0.conceptId, $0) })
        return row.concepts.map { concept in
            let cell = cellsByConcept[concept.id]
            return MasteryConceptRow(
                id: concept.id,
                name: concept.name,
                score: cell?.masteryScore,
                assessed: cell?.assessed ?? false,
                level: level(score: cell?.masteryScore, assessed: cell?.assessed ?? false)
            )
        }
        .sorted { lhs, rhs in
            if lhs.assessed != rhs.assessed { return !lhs.assessed && rhs.assessed }
            return (lhs.score ?? 0) < (rhs.score ?? 0)
        }
    }

    static func summary(_ rows: [MasteryConceptRow]) -> (mastered: Int, atRisk: Int, total: Int) {
        let assessed = rows.filter(\.assessed)
        return (
            mastered: assessed.filter { $0.level == .mastered }.count,
            atRisk: assessed.filter { $0.level == .atRisk }.count,
            total: rows.count
        )
    }

    static func cacheKeyMastery(courseCode: String, enrollmentId: String) -> String {
        "mastery:\(courseCode):\(enrollmentId)"
    }

    static func cacheKeyMyReportCards() -> String {
        "mastery:my-report-cards"
    }

    static func releasedReportCards(_ cards: [ReportCardSummary]) -> [ReportCardSummary] {
        cards.filter { $0.status == "released" }
            .sorted { $0.gradingPeriod > $1.gradingPeriod }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/MasteryLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.R

enum class MasteryLevel {
    MASTERED,
    DEVELOPING,
    BEGINNING,
    AT_RISK,
    NOT_ASSESSED,
}

data class MasteryConceptRow(
    val id: String,
    val name: String,
    val score: Double?,
    val assessed: Boolean,
    val level: MasteryLevel,
)

object MasteryLogic {
    /** Matches web `masteryLabel()` thresholds (0.8 / 0.6 / 0.4). */
    fun level(score: Double?, assessed: Boolean): MasteryLevel {
        if (!assessed || score == null) return MasteryLevel.NOT_ASSESSED
        return when {
            score >= 0.8 -> MasteryLevel.MASTERED
            score >= 0.6 -> MasteryLevel.DEVELOPING
            score >= 0.4 -> MasteryLevel.BEGINNING
            else -> MasteryLevel.AT_RISK
        }
    }

    fun levelLabelRes(level: MasteryLevel): Int = when (level) {
        MasteryLevel.MASTERED -> R.string.mobile_mastery_levelMastered
        MasteryLevel.DEVELOPING -> R.string.mobile_mastery_levelDeveloping
        MasteryLevel.BEGINNING -> R.string.mobile_mastery_levelBeginning
        MasteryLevel.AT_RISK -> R.string.mobile_mastery_levelAtRisk
        MasteryLevel.NOT_ASSESSED -> R.string.mobile_mastery_levelNotAssessed
    }

    fun rows(row: StudentMasteryRow): List<MasteryConceptRow> {
        val cellsByConcept = row.cells.associateBy { it.conceptId }
        return row.concepts
            .map { concept ->
                val cell = cellsByConcept[concept.id]
                MasteryConceptRow(
                    id = concept.id,
                    name = concept.name,
                    score = cell?.masteryScore,
                    assessed = cell?.assessed ?: false,
                    level = level(cell?.masteryScore, cell?.assessed ?: false),
                )
            }
            .sortedWith(
                compareBy(
                    { it.assessed },
                    { it.score ?: 0.0 },
                ),
            )
    }

    data class Summary(val mastered: Int, val atRisk: Int, val total: Int)

    fun summary(rows: List<MasteryConceptRow>): Summary {
        val assessed = rows.filter { it.assessed }
        return Summary(
            mastered = assessed.count { it.level == MasteryLevel.MASTERED },
            atRisk = assessed.count { it.level == MasteryLevel.AT_RISK },
            total = rows.size,
        )
    }

    fun cacheKeyMastery(courseCode: String, enrollmentId: String): String =
        "mastery:$courseCode:$enrollmentId"

    fun cacheKeyMyReportCards(): String = "mastery:my-report-cards"

    fun releasedReportCards(cards: List<ReportCardSummary>): List<ReportCardSummary> =
        cards.filter { it.status == "released" }.sortedByDescending { it.gradingPeriod }
}
